File: Recursion/leetcode_241/solution.js
/**
 * @param {string} expression
 * @return {number[]}
 */
var diffWaysToCompute = function(expression) {
    const memo = new Map();

    function compute(subExpression) {
        if (memo.has(subExpression)) {
            return memo.get(subExpression);
        }

        const result = [];
        let isNumber = true;

        for (let i=0; i<subExpression.length; i++) {
            const char = subExpression[i];

            if (char === '+' || char === '-' || char === '*') {
                isNumber = false;
                
                const leftResult = compute(subExpression.substring(0, i));
                const rightResult = compute(subExpression.substring(i + 1));

                for (const leftVal of leftResult) {
                    for (const rightVal of rightResult) {
                        if (char === '+') {
                            result.push(leftVal + rightVal);
                        } else if (char === '-') {
                            result.push(leftVal - rightVal);
                        } else {
                            result.push(leftVal * rightVal);
                        }
                    }
                }
            }
        }

        if (isNumber) {
            result.push(Number(subExpression));
        }

        memo.set(subExpression, result);
        return result;
    }

    return compute(expression);
};

File: Recursion/leetcode_241/solution.py
from typing import List

class Solution:
    def diffWaysToCompute(self, expression: str) -> List[int]:
        self.memo = {}

        def compute(sub_expression: str) -> list[int]:
            if sub_expression in self.memo:
                return self.memo[sub_expression]

            result = []
            is_number = True

            for i, char in enumerate(sub_expression):
                if char in ['+', '-', '*']:
                    is_number = False

                    left_result = compute(sub_expression[:i])
                    right_result = compute(sub_expression[i+1:])

                    for left_val in left_result:
                        for right_val in right_result:
                            if char == '+':
                                result.append(left_val + right_val)
                            elif char == '-':
                                result.append(left_val - right_val)
                            else:
                                result.append(left_val * right_val)
            
            if is_number:
                result.append(int(sub_expression))

            self.memo[sub_expression] = result
            return result
        
        return compute(expression)